// Locations/saveLocation.js
var middle = require('./init');
var location = require('./locations');
var assert = require('assert');
var dateFormat = require('dateformat');
var mongoose = require('mongoose');
exports.saveCoordinations = function(token, latitude, longidude, callback) {
  var now = new Date();
  var month = now.getUTCMonth()+1; //months from 1-12
  var day = now.getUTCDate();
  var year = now.getUTCFullYear();
  var date = day + "/" + month + "/" + year;
  console.log(date);

  var newLocation = new location ({
    token : token,
    latitude : latitude,
    longidude : longidude,
    date : date
  });

  newLocation.save();

  callback({'response':"Location Saved",'res':true});
}

exports.getCoordinations = function(token,callback){

  location.find({ 'token' : token}, 'latitude longidude', function(err,locations) {
    if(err){
      callback({'response': 'This token does not exists','res':false});
    } else {
      callback({'response': locations,'res':true});
    }
  });
}

exports.getCoordinationsDates = function(token,dates, callback){
  console.log('log: ' + dates);
  location.find({'token' : token , 'date' : dates},'latitude longidude', function(err,locations){
    if(err){
      callback({'response':'Nothing to display','res':false});
    } else {
      callback({'response':locations,'res':true});
    }
  });
}
